### PR TITLE
Add metric -- count of queries waiting for merge buffers

### PR DESCRIPTION
**Pull request by kaisun2000**
### Description

Currently there is no metric to indicate the contention level for the shared resource -- merge buffer pools. Thus, it is hard to know that the query performance suffers from the contention for shared merge buffer pools. We did meet such performance degradation issue in the production environment. 

This PR exposed the count of waiting queries (threads) blocking at in the merge buffers pools. 

Summary of the changes for a quick reference:

- `BlockingPool` interface was updated with a new method `getPendingQueries` to return the count of pending queries.
-  `DefaultBlockingPool` class was updated with the implementation of gathering the number of pending queries with an `AtomicLong` counter.
-  `MergeBufferPoolMonitor` class was created with the shared merge buffer pool (annotated with `@Merging`) injected.  User can add this monitor to the configuration file such as `druid.monitoring.monitors=["org.apache.druid.java.util.metrics.MergeBufferPoolMonitor", ...]`
- Unit test is added. This unit test validated the `MergeBufferPoolMonitor` would emit correct count of pending queries with a query blocking at the `DefaultBlockingPool` by taking more than the number of available buffers in the `DefaultBlockingPool`

#### Release note

<hr>

##### Key changed/added classes in this PR
 * You can monitor the number of pending queries at the merge buffer pool with the metric `mergebuffer/pendingQueries`
 * Add the MergeBufferPoolMonitor to the configuration files of Broker, Historical or Router as `druid.monitoring.monitors=["org.apache.druid.java.util.metrics.MergeBufferPoolMonitor", ...]`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java [hunk 1]
```diff
@@ -48,6 +49,9 @@ public class DefaultBlockingPool<T> implements BlockingPool<T>
   private final Condition notEnough;
   private final int maxSize;
 
+  private final AtomicLong pendingQueries;
+
+
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:54]:
```suggestion
```

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java [hunk 2]
```diff
@@ -62,6 +66,8 @@ public DefaultBlockingPool(
 
     this.lock = new ReentrantLock();
     this.notEnough = lock.newCondition();
+
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:69]:
```suggestion
```

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java [hunk 3]
```diff
@@ -91,26 +97,38 @@ public List<ReferenceCountingResourceHolder<T>> takeBatch(final int elementNum,
...
   @Override
   public List<ReferenceCountingResourceHolder<T>> takeBatch(final int elementNum)
   {
     checkInitialized();
     try {
+      pendingQueries.incrementAndGet();
       return takeObjects(elementNum).stream().map(this::wrapObject).collect(Collectors.toList());
     }
     catch (InterruptedException e) {
+      pendingQueries.incrementAndGet();
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:121]: Seems like this is missing a finally block to decrement pendingQueries
```suggestion
```

**Review by suneet-s — COMMENTED** [review]
Thanks for your first contribution!

I took a quick pass of the code and have some suggestions - specifically in the `MergeBufferPoolMonitor` that should make it easier to deploy.

Since this monitor runs only once on every emission period, the metrics only provide a snapshot in time of how many queries were waiting on merge buffers. Have you considered reporting metrics like mean, P98 and max time spent waiting on merge buffers instead?

That should provide visibility into how long queries are taking to acquire merge buffers on all the different services.

Another metric that may be interesting is the number of queries that timed out while waiting to acquire merge buffers.

**Review by LakshSingla — COMMENTED** [review]
Thanks for the first PR! Left a few review comments. 
Also, we should add the monitor to `docs/configuration/index.md` along with the other monitors.

**Comment by abhishekagarwal87** [comment]
We don't need to add another monitor. The new metrics can be emitted from the QueryCountStatsMonitor.

**Comment by kaisun2000** [comment]
> We don't need to add another monitor. The new metrics can be emitted from the QueryCountStatsMonitor.

@abhishekagarwal87, just want to pan out the overall structure of add this metric from the `QueryCountStatsMonitor` and confirm the design before I make the changes. 

`QueryCountStatsMonitor` takes a `QueryCountStatsProvider` in its constructor to provide the QueryCountStats (aggregate states instead of query specific stats). The `QueryCountStatsProvider` is implemented as `QueryResource` (which implements the logic for query REST endpoints). 

So here, I can add a parameter in `QueryResource` constructor as `@Merging BlockingPool<ByteBuffer> mergeBufferPoolIn`. And let Guice to inject the shared merge buffer pool. This way, we can reuse the same monitor.

In the interface of `QueryCountStatsProvider`, we need to add a method to report the pending query count such as `
  public long getPendingQueriesCount()`
  
We will also need to keep the same changes in this PR to add a method in `BlockingPool` to report the pending queries in the pool, as `public long getPendingQueries()`
 
 Let me know if I missed anything. Or next I will update the PR to reflect this changes discussed here.

File: processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import java.nio.ByteBuffer;
+
+public class MergeBufferPoolMonitor extends AbstractMonitor
+{
+
+  private final BlockingPool<ByteBuffer> mergeBufferPool;
+
+  @Inject
+  public MergeBufferPoolMonitor(
+      @Merging BlockingPool<ByteBuffer> mergeBufferPoolIn
```
[the remark this answers is not in the document]
Comment (abhishekagarwal87) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java:37]: @LakshSingla is there a reason you want to do it that way? what structure do you have in mind exactly?

**Comment by abhishekagarwal87** [comment]
@kaisun2000 - How about injecting BlockingPool directly into the QueryCountStatsMonitor? Though seems like Laksh has objections to injecting pool into a monitor. I will let him chime in too.

**Comment by kaisun2000** [comment]
> @kaisun2000 - How about injecting BlockingPool directly into the QueryCountStatsMonitor? Though seems like Laksh has objections to injecting pool into a monitor. I will let him chime in too.

@abhishekagarwal87, thx for coordinating this and reached an consensus. 

In this case, I will inject the `BlockinPool` to `QueryCountStatsMonitor` and not touching the `QueryCountStatsProvider` interface. Will send the revised PR next.

File: processing/src/main/java/org/apache/druid/collections/BlockingPool.java [hunk 5]
```diff
@@ -40,8 +39,14 @@ public interface BlockingPool<T>
    * Take resources from the pool, waiting if necessary until the elements of the given number become available.
    *
    * @param elementNum number of resources to take
-   *
    * @return a list of resource holders. An empty list is returned if {@code elementNum} resources aren't available.
    */
   List<ReferenceCountingResourceHolder<T>> takeBatch(int elementNum);
+
+  /**
+   * Return the count of pending queries blocking in this queue for merge buffers
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/collections/BlockingPool.java:47]: change.

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java (hunk 1 above)
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:54]: deleted.

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java (hunk 2 above)
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:69]: deleted.

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java (hunk 3 above)
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:121]: updated. Good catch

File: processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.google.inject.Inject;
+import org.apache.druid.collections.BlockingPool;
+import org.apache.druid.guice.annotations.Merging;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+
+import java.nio.ByteBuffer;
+
+public class MergeBufferPoolMonitor extends AbstractMonitor
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java:30]: Not applied anymore as we are removing this monitor.

File: processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java]: good point. Not applied anymore as moving the logic to the `QueryStatsCountMonitor` as @abhishekagarwal87 suggested

File: processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  {
+    this.mergeBufferPool = mergeBufferPoolIn;
+  }
+
+  @Override
+  public boolean doMonitor(ServiceEmitter emitter)
+  {
+    long pendingQueries = this.mergeBufferPool.getPendingQueries();
+
+    ServiceMetricEvent.Builder builder = ServiceMetricEvent.builder().setFeed("metrics");
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java]: changed as suggested.

File: processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    emitter.emit(builder.setMetric("mergebuffer/pendingQueries", pendingQueries));
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java]: added as suggested.

[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitor.java]: changed as suggested.

File: processing/src/test/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitorTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    });
+
+    try {
+      // the latch returns from await() guarantees the above lamda to take buffer from the pool starting to run in the
+      // executorService thread
+      latch.await();
+
+      // give 1 sec for pool.takeBatch to run and blocking at the pool
+      Thread.sleep(1000);
+
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/test/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitorTest.java]: Good point. In the new test implementation, we do a check every 0.1 sec and limit the waiting till 10 secs at most. This way, we eliminated the theoretical potential race condition (relying on sleep). And also give the test a finishing time bound.

File: processing/src/test/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitorTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      latch.countDown();
+      pool.takeBatch(10);
+    });
+
+    try {
+      // the latch returns from await() guarantees the above lamda to take buffer from the pool starting to run in the
+      // executorService thread
+      latch.await();
+
+      // give 1 sec for pool.takeBatch to run and blocking at the pool
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/test/java/org/apache/druid/java/util/metrics/MergeBufferPoolMonitorTest.java]: changed as suggested.

**Comment by LakshSingla** [comment]
> Though seems like Laksh has objections to injecting pool into a monitor. I will let him chime in too.

I had concerns with the monitor directly interacting with the merge buffer pool, instead of the stats provider that it was supposed to work with. However, I don't have a strong stance on the same, since its mostly internal implementation. The changes.

File: processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java [hunk 11]
```diff
@@ -48,6 +49,8 @@ public class DefaultBlockingPool<T> implements BlockingPool<T>
   private final Condition notEnough;
   private final int maxSize;
 
+  private final AtomicLong pendingQueries;
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/collections/DefaultBlockingPool.java:52]: nit: refactor as above

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java [hunk 12]
```diff
@@ -21,24 +21,30 @@
...
 {
   private final KeyedDiff keyedDiff = new KeyedDiff();
   private final QueryCountStatsProvider statsProvider;
+  private final BlockingPool<ByteBuffer> mergeBufferPool;
 
   @Inject
   public QueryCountStatsMonitor(
-      QueryCountStatsProvider statsProvider
+      QueryCountStatsProvider statsProvider,
+      @Merging BlockingPool<ByteBuffer> mergeBufferPoolIn
```
Comment (LakshSingla) [inline server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:43]: Why is it named `mergeBufferPoolIn`?

**Review by LakshSingla — COMMENTED** [review]
Overall LGTM left a few comments. TY for taking this up!

File: server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java (hunk 12 above)
Comment (kaisun2000) [reply server/src/main/java/org/apache/druid/server/metrics/QueryCountStatsMonitor.java:43]: to follow other code patterns, changed to 

```
  @Inject
  public QueryCountStatsMonitor(
      QueryCountStatsProvider statsProvider,
      @Merging BlockingPool<ByteBuffer> mergeBufferPool
  )
  {
    this.statsProvider = statsProvider;
    this.mergeBufferPool = mergeBufferPool;
  }

```

File: server/src/test/java/org/apache/druid/server/metrics/QueryCountStatsMonitorTest.java [hunk 13]
```diff
@@ -85,12 +105,45 @@ public void testMonitor()
...
+  @Test
+  public void testMonitoringMergeBuffer()
+  {
+    executorService.submit(() -> {
+      mergeBufferPool.takeBatch(10);
+    });
+
+    int count = 0;
+    try {
+      // wait at most 10 secs for the executor thread to block
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply server/src/test/java/org/apache/druid/server/metrics/QueryCountStatsMonitorTest.java:127]: added `@Test(timeout = 2_000L)` as suggested.

File: processing/src/main/java/org/apache/druid/collections/BlockingPool.java [hunk 14]
```diff
@@ -40,8 +39,14 @@ public interface BlockingPool<T>
...
    * @return a list of resource holders. An empty list is returned if {@code elementNum} resources aren't available.
    */
   List<ReferenceCountingResourceHolder<T>> takeBatch(int elementNum);
+
+  /**
+   * Return the count of pending queries waiting for merge buffers
+   *
+   * @return count of pending queries
+   */
+  long getPendingQueries();
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply processing/src/main/java/org/apache/druid/collections/BlockingPool.java]: updated as suggested.

**Review by LakshSingla — APPROVED** [review]
Code coverage issues, however, the coverage looks good to me. The insufficiency is due to the `DummyBlockingPool` not being tested anywhere, which is okay to me. 

LGTM 🚀

**Comment by LakshSingla** [comment]
ty for the contribution, I am merging the PR. Can you please add a release note to the description @kaisun2000? (a few line user-facing summary of the change) It helps in prepping up the release notes and calling out user-facing changes.

**Comment by kaisun2000** [comment]
@LakshSingla, just added the release notes section. Thx for help to get this PR reviewed and checked in.

**Comment by abhishekagarwal87** [comment]
thank you for your contribution @kaisun2000

**Pull request merged**